rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+---
+authors: Krzysztof Skrzętnicki (krzysztof.skrzetnicki@goteleport.com)
+state: draft
+---
+
+# RFD 151 - Database Permission Management
+
+## Required Approvals
+
+- Engineering: @r0mant && @smallinsky
+- Product: @xinding33 || @klizhentas
+- Security: @reedloden || @jentfoo
+
+## What
+
+Extends [automated db user provisioning](0113-automatic-database-users.md) with
+permission management capabilities.
+
+## Why
+
+Managing database-level permissions is a natural extension of current Teleport
+RBAC capabilities. The described model should also integrate seamlessly with
+TAG, providing a future-proof solution.
+
+Database administrators will be able to use Teleport for both user and
+permission management.
+
+## Details
+
+### UX
+
+A set of global "database object import rules" resources are defined for the
+Teleport cluster. At certain points in time, the database schema is read and
+passed through the import rules. The individual import rules may apply custom
+labels. An database object resource will only be created if there is at least
+one import rule that matched it.
+
+Only object attributes (standard ones like `protocol` or custom from the
+`attributes` field), which are sourced from the object spec, are subject to
+matching against the import rules. Any labels present on an object, either from
+an import rule or another source, are not subject to be matched by the import
+rule. The permissions matching is different in this regard.
+
+Example: the `sales-prod` import rule, which imports all tables from `sales`
+Postgres database in prod:
+
+```yaml
+kind: db_object_import_rule
+metadata:
+  name: sales-prod
+spec:
+  db_labels:
+    env: prod
+  mappings:
+    - object_match:
+        - database: sales
+          object_kind: table
+          protocol: postgres
```

Comment:
Why is the "protocol" needed here? I'm thinking if you need to create an import rule for, say, Postgres databases, you mark them with appropriate label and filter here through "db_labels" above?

Reply:
TBH I still think that having the protocol here is a little confusing and that labels are flexible enough for users to be able to construct rules they want. I would keep it out of initial scope, we can always add it later if needed.

Also I agree that we should support table names, something similar to Okta import rules like:

```yaml
match:
- database_names: ["pii", "db*"]
  table_names: ["customers", "products"]
```

Reply:
I went ahead and simplified the permissions matching part; now it is only done via labels, and not object attributes (name, schema, ...). You can still get the same effective behaviour if you add an import rule which propagates the attributes that you need (which is some work), but the positive part is that the implementation gets easier to write and understand; the latter part is a win in my book, as it makes the UX better. The roles are shorter, too.